File: src/2015/10/day.10.py
from timeit import default_timer as timer
from typing import Tuple


def parse_input():
    with open("src/inputs/2015/10/day.10.input.txt", "r") as file:
        return file.readline()


def part1():
    sequence = parse_input()

    for _ in range(40):
        sequence = look_and_say(sequence)

    return len(sequence)


def part2():
    sequence = parse_input()

    for _ in range(50):
        sequence = look_and_say(sequence)

    return len(sequence)


def get_runs(value: str) -> list[Tuple[str, int]]:
    value = value + "~"

    sequence = []

    run_character = value[0]
    run_count = 0

    for character in value:
        if character != run_character:
            sequence.append((run_character, run_count))
            run_count = 0

        run_character = character
        run_count += 1

    return sequence


def look_and_say(sequence: str) -> str:
    runs = get_runs(sequence)

    new_sequence = ""

    for run in runs:
        new_sequence += str(run[1]) + run[0]

    return new_sequence


def timed(fn):
    start = timer()
    result = fn()

    return (timer() - start, result)


(p1_elapsed, p1_answer) = timed(part1)
print(f"Part 1: {p1_answer} ({p1_elapsed:.3f}s)")

(p2_elapsed, p2_answer) = timed(part2)
print(f"Part 2: {p2_answer} ({p2_elapsed:.3f}s)")

File: src/2015/10/day.10.ts
import fs from "fs";

function parseInput() {
  return fs.readFileSync("src/inputs/2015/10/day.10.input.txt").toString();
}

export function part1() {
  return lookAndSay(parseInput(), 40).length;
}

export function part2() {
  return lookAndSay(parseInput(), 50).length;
}

function lookAndSay(sequence: string, iterations: number) {
  for (let n = 0; n < iterations; n++) {
    sequence += "~";
    let sequence_ = "";

    let i = 0;

    let runDigit = sequence[i];
    let runCount = 0;

    while (i < sequence.length) {
      if (sequence[i] !== runDigit) {
        sequence_ += String(runCount) + runDigit;
        runCount = 0;
      }

      runDigit = sequence[i];
      runCount++;

      i++;
    }

    sequence = sequence_;
  }

  return sequence;
}
